### PR TITLE
Provide Python 3.8.0a3

### PR DESCRIPTION
**Pull request by h-vetinari**
First (and likely biggest) step towards #254

Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)

Initial work based on [comment](https://github.com/conda-forge/python-feedstock/issues/254#issuecomment-484353300) by @jakirkham. I'm guessing most patches should be reviewed if they are still necessary? Or they'll probably just fail the rerender? Let's see... :)

### CODE REVIEW

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
As somewhat expected, applying the patches fails. Will try to have a look later. Is there any way to see the output from the failed azure pipelines, or do I have to test the patches locally?

**Comment by jjhelmus** [comment]
> As somewhat expected, applying the patches fails. Will try to have a look later. Is there any way to see the output from the failed azure pipelines, or do I have to test the patches locally?

I updated the patches and disabled profiled guided optimization which will speed up the builds.  

You can see the output of the azure pipelines if you click on the "Details" link and then "View more details on Azure Pipelines".

**Comment by jjhelmus** [comment]
We will want to push this to a different branch and a label, any suggestions? Perhaps `pre-3.8` for the branch and re-use the `prerelease` label on Anaconda Cloud.

**Comment by h-vetinari** [comment]
> @jjhelmus: I updated the patches and disabled profiled guided optimization which will speed up the builds.

Thanks for taking this up, and sorry for the late response - I was AFK over the holidays.

> You can see the output of the azure pipelines if you click on the "Details" link and then "View more details on Azure Pipelines".

Thanks, I knew that, but I thought there would be a richer output for the git apply fail. In any case, you already solved the patches.

The last remaining failure is about a casting error for a version number:
```
Build FAILED.

       "%SRC_DIR%\PCbuild\pcbuild.proj" (Build target) (1) ->
       "%SRC_DIR%\PCbuild\pythoncore.vcxproj" (Build target) (2) ->
         %SRC_DIR%\PCbuild\python.props(94,31): error MSB4086: A numeric comparison was attempted on "$([System.Version]::Parse($(_RegistryVersion)))" that evaluates to "10.0.17763.0" instead of a number, in condition "$([System.Version]::Parse($(_RegistryVersion))) > $([System.Version]::Parse($(DefaultWindowsSDKVersion)))". [%SRC_DIR%\PCbuild\pythoncore.vcxproj]

    0 Warning(s)
    1 Error(s)
```

After a bit of googling, this seems to be due to the MSBuild version (see for example [here](https://github.com/Microsoft/msbuild/issues/778), esp. the closing comment). Any reason to not use a newer one for the Azure Build?

> We will want to push this to a different branch and a label, any suggestions? Perhaps `pre-3.8` for the branch and re-use the `prerelease` label on Anaconda Cloud.

That would certainly be a short-term option. But maybe it's also time to discuss installing pre-releases with conda again...? xref https://github.com/conda-forge/conda-forge.github.io/issues/147 https://github.com/conda/conda/issues/3125

File: recipe/patches/0017-Unvendor-openssl.patch [hunk 1]
```diff
@@ -71,33 +71,33 @@ index f83d4df0d5..b2a9f798f2 100644
...
    <ItemGroup>
      <ClCompile Include="..\Programs\python.c" />
-+    <ClCompile Include="$(opensslIncludeDir)\openssl\applink.c">
-+      <PreprocessorDefinitions>_CRT_SECURE_NO_WARNINGS;$(PreprocessorDefinitions)</PreprocessorDefinitions>
-+    </ClCompile>
++    <ClCompile Include="$(opensslIncludeDir)\openssl\applink.c">

++      <PreprocessorDefinitions>_CRT_SECURE_NO_WARNINGS;$(PreprocessorDefinitions)</PreprocessorDefinitions>

++    </ClCompile>
```
Comment (h-vetinari) [inline recipe/patches/0017-Unvendor-openssl.patch]: Line ending artifacts...?

File: recipe/patches/0017-Unvendor-openssl.patch [hunk 2]
```diff
@@ -71,33 +71,33 @@ index f83d4df0d5..b2a9f798f2 100644
...
    <ItemGroup>
      <ClCompile Include="..\PC\WinMain.c" />
-+    <ClCompile Include="$(opensslIncludeDir)\openssl\applink.c">
-+      <PreprocessorDefinitions>_CRT_SECURE_NO_WARNINGS;$(PreprocessorDefinitions)</PreprocessorDefinitions>
-+    </ClCompile>
++    <ClCompile Include="$(opensslIncludeDir)\openssl\applink.c">

++      <PreprocessorDefinitions>_CRT_SECURE_NO_WARNINGS;$(PreprocessorDefinitions)</PreprocessorDefinitions>

++    </ClCompile>
```
Comment (h-vetinari) [inline recipe/patches/0017-Unvendor-openssl.patch]: Line ending artifacts...?

File: recipe/patches/0018-Unvendor-sqlite3.patch [hunk 3]
```diff
@@ -11,24 +11,24 @@ Subject: [PATCH 18/18] Unvendor sqlite3
...
        <PreprocessorDefinitions>MODULE_NAME="sqlite3";%(PreprocessorDefinitions)</PreprocessorDefinitions>
      </ClCompile>
-+    <Link>
-+      <AdditionalDependencies>$(sqlite3Dir)\lib\sqlite3.lib;%(AdditionalDependencies)</AdditionalDependencies>
-+    </Link>
++    <Link>

++      <AdditionalDependencies>$(sqlite3Dir)\lib\sqlite3.lib;%(AdditionalDependencies)</AdditionalDependencies>

++    </Link>
```
Comment (h-vetinari) [inline recipe/patches/0018-Unvendor-sqlite3.patch]: Line ending artifacts...?

File: recipe/patches/0018-Unvendor-sqlite3.patch [hunk 4]
```diff
@@ -79,15 +79,15 @@ index 4f5b1965d9..59cae10127 100644
...

++    <SqliteVersion>$(_SqliteVersion)</SqliteVersion>

++    <SqliteMajorVersion>$(_SqliteVersion.Split(`.`)[1])</SqliteMajorVersion>

++    <SqliteMinorVersion>$(_SqliteVersion.Split(`.`)[2])</SqliteMinorVersion>

++    <SqliteMicroVersion>$(_SqliteVersion.Split(`.`)[3])</SqliteMicroVersion>

++    <SqlitePatchVersion>$(_SqliteVersion.Split(`.`)[4])</SqlitePatchVersion>
```
Comment (h-vetinari) [inline recipe/patches/0018-Unvendor-sqlite3.patch]: Line ending artifacts...?

File: recipe/patches/0017-Unvendor-openssl.patch (hunk 1 above)
Comment (jjhelmus) [reply recipe/patches/0017-Unvendor-openssl.patch]: The CRLF line endings added in this patch are correct.  This is changing a Windows file which uses these line endings.  The previous patch had incorrect line endings.

**Comment by jjhelmus** [comment]
> After a bit of googling, this seems to be due to the MSBuild version (see for example [here](https://github.com/Microsoft/msbuild/issues/778), esp. the closing comment). Any reason to not use a newer one for the Azure Build?

The Azure Windows images have some issue building certain software, something to do with how MSBuild is configured or installed.  @mariusvniekerk may know how.  In cases like these AppVeyor is used for this builds and in this cases is passing.

**Comment by h-vetinari** [comment]
> @jjhelmus: In cases like these AppVeyor is used for this builds and in this cases is passing.

Does it mean that this recipe is essentially ready then?

**Comment by jjhelmus** [comment]
> Does it mean that this recipe is essentially ready then?

Yes the recipe should now be ready.  I targeted the PR against a new "pre-3.8" branch and adjusted to `conda-forge.yml` file to uploaded the packages to the `pre-3.8` label.  Once the CI passes I'll merge the PR and soon afterwards the Python 3.8.0a3 packages will be available.

**Comment by jjhelmus** [comment]
Oops, changing the upload target in 00fa649 did have any effect due to conda-forge/conda-smithy#897.  No a big deal, I will set the labels manually on Anaconda Cloud.

**Pull request merged**